### ISSUE TITLE
tracer example goes off track at getenv

### ISSUE THREAD

**Issue by bannsec** [angr/archr#27]
Following your tracer test case for archer, it seems to setup correctly and trace part way into the libc. It then errors out:

```
In [18]: simgr.run()
---------------------------------------------------------------------------
TracerDesyncError                         Traceback (most recent call last)
<ipython-input-18-d7a0b3e79488> in <module>
----> 1 simgr.run()

~/angr-dev/angr/angr/sim_manager.py in run(self, stash, n, until, **kwargs)
    257         for _ in (itertools.count() if n is None else range(0, n)):
    258             if not self.complete() and self._stashes[stash]:
--> 259                 self.step(stash=stash, **kwargs)
    260                 if not (until and until(self)):
    261                     continue

~/angr-dev/angr/angr/misc/hookset.py in __call__(self, *args, **kwargs)
     73             current_hook = self.pending.pop()
     74             try:
---> 75                 result = current_hook(self.func.__self__, *args, **kwargs)
     76             finally:
     77                 self.pending.append(current_hook)

~/angr-dev/angr/angr/exploration_techniques/tracer.py in step(self, simgr, stash, **kwargs)
    223     def step(self, simgr, stash='active', **kwargs):
    224         simgr.drop(stash='missed')
--> 225         return simgr.step(stash=stash, **kwargs)
    226
    227     def step_state(self, simgr, state, **kwargs):

~/angr-dev/angr/angr/misc/hookset.py in __call__(self, *args, **kwargs)
     78             return result
     79         else:
---> 80             return self.func(*args, **kwargs)

~/angr-dev/angr/angr/sim_manager.py in step(self, stash, n, selector_func, step_func, successor_func, until, filter_func, **run_args)
    341
    342             pre_errored = len(self._errored)
--> 343             successors = self.step_state(state, successor_func=successor_func, **run_args)
    344
    345             # handle degenerate stepping cases here. desired behavior:

~/angr-dev/angr/angr/misc/hookset.py in __call__(self, *args, **kwargs)
     73             current_hook = self.pending.pop()
     74             try:
---> 75                 result = current_hook(self.func.__self__, *args, **kwargs)
     76             finally:
     77                 self.pending.append(current_hook)

~/angr-dev/angr/angr/exploration_techniques/tracer.py in step_state(self, simgr, state, **kwargs)
    273             # strict mode
    274             if len(succs) == 1:
--> 275                 self._update_state_tracking(succs[0])
    276             elif len(succs) == 0:
    277                 raise Exception("All states disappeared!")

~/angr-dev/angr/angr/exploration_techniques/tracer.py in _update_state_tracking(self, state)
    438             raise TracerDesyncError("Oops! angr did not follow the trace",
    439                                     deviating_addr=state.addr,
--> 440                                     deviating_trace_idx=idx+1)
    441
    442         if state.globals['sync_idx'] is not None:

TracerDesyncError: Oops! angr did not follow the trace
```

Looking at the state it went offline at:
```
 '<IRSB from 0x7ffff7a26778: 1 sat>',
 '<IRSB from 0x7ffff7a26785: 1 sat>',
 '<IRSB from 0x7ffff7a26778: 1 sat>',
 '<IRSB from 0x7ffff7a26785: 1 sat>',
 '<IRSB from 0x7ffff7a26778: 1 sat>',
 '<IRSB from 0x7ffff7a26785: 1 sat>',
 '<IRSB from 0x7ffff7a26778: 1 sat>']

In [23]: p
Out[23]: <Project /home/angr/work/ctf/gctf/2019/rev/flaggy_bird/cat>

In [24]: p.loader.describe_addr(0x7ffff7a26778)
Out[24]: 'getenv+0x98 in libc.so.6 (0x42778)'
```

Resolving what `getenv` string it was attempting to look at
```
In [38]: s2.mem[s2.regs.r13].string.concrete
Out[38]: b'CPATH'
```

CPATH does not appear to exist in the environment given to angr (nor the base env). Not sure at this point why it's diverging from the QEMU trace.

**Comment by rhelmot** [comment]
as usual, gonna need a testcase to actually examine this. You might want to check that that pointer actually points into the area being used to store the environment variables. It might be the value the binary is trying to lookup, and the desync is that there's just a different number of environment variables between the two runs, even though they would both come back with "not found". That would still be a bug, though.

**Comment by bannsec** [comment]
Can upload in a bit, but test binary is just cat from ubuntu:bionic

**Comment by rhelmot** [comment]
Okay, I was able to reproduce just by copying my own cat (I am on bionic) and running the test_angr_tracing testcase. I also verified that the environment found by the datascout is identical to the environment present in memory, so there's a problem somewhere in the datascout bow, or with our assumptions about how the environment will be manipulated between different processes.

Gonna assign this to yan, since this was his child.
